calyx-backend/src/firrtl.rs

```diff
@@ -84,41 +92,106 @@ fn emit_component<F: io::Write>(
     // Add a COMPONENT START: <name> anchor before any code in the component
     writeln!(f, "{}; COMPONENT START: {}", SPACING.repeat(2), comp.name)?;
 
-    // TODO: Cells. NOTE: leaving this one for last
+    // Cells
+    for cell in comp.cells.iter() {
+        let cell_borrowed = cell.as_ref().borrow();
+        if cell_borrowed.type_name().is_some() {
+            match cell_borrowed.prototype {
+                ir::CellType::Primitive {
+                    name: _,
+                    param_binding: _,
+                    is_comb: _,
+                    latency: _,
+                } => {
+                    // TODO: use extmodules
+                    writeln!(
+                    f,
+                    "{}; FIXME: attempting to instantiate primitive cell {}",
+                    SPACING.repeat(2),
+                    cell_borrowed.name()
+                )?
+                }
+                ir::CellType::Component { name } => {
+                    writeln!(
+                        f,
+                        "{}inst {} of {}",
+                        SPACING.repeat(2),
+                        cell_borrowed.name(),
+                        name
+                    )?;
+                }
+                ir::CellType::ThisComponent => unreachable!(),
+                ir::CellType::Constant { val: _, width: _ } => unreachable!(),
+            }
+        }
+    }
 
+    let mut dst_set: HashSet<String> = HashSet::new();
+    // Emit assignments
     for asgn in &comp.continuous_assignments {
-        // TODO: guards
         match asgn.guard.as_ref() {
-            ir::Guard::Or(_, _) => todo!(),
-            ir::Guard::And(_, _) => todo!(),
-            ir::Guard::Not(_) => todo!(),
             ir::Guard::True => {
                 // Simple assignment with no guard
-                let _ = write_assignment(asgn, f);
+                let _ = write_assignment(asgn, f, 2);
+            }
+            _ => {
+                let dst_canonical = &asgn.dst.as_ref().borrow().canonical();
+                let dst_canonical_str = dst_canonical.to_string();
+                if !dst_set.contains(&dst_canonical_str) {
+                    // if we don't have a "is invalid" statement yet, then we have to write one.
+                    let _ = write_invalid_initialization(&asgn.dst, f);
+                    dst_set.insert(dst_canonical_str);
+                }
```

Comment:
This uses a "lazy" policy to initialize all ports that are assigned to at least once. Makes sense and no need to change it, but I just wanted to mention that we could consider an "eager" policy as well. Basically, we'd enumerate _all possible ports_ (which would consist of all our children's input ports and our output ports) and do an `is invalid` to all of them up front.

This isn't necessarily any better than the lazy approach, except maybe that it doesn't require the `dst_set` tracking. So let's not do it, but I thought I'd record the idea in case it becomes relevant in the future.

Reply:
Ah that makes sense!! Let me leave a comment in the code so that I don't forget as well :)